File: 008/string-to-integer.js
/**
 * @param {string} str
 * @return {number}
 */
var myAtoi = function (str) {
    str = str.trim();
    let num = 0;

    let INT_MIN = 1 << 31;
    let INT_MAX = -(1 << 31) - 1;
    let pattern = /^(\-|\+)?[0-9]+/;
    let tmp = pattern.exec(str);

    if (tmp) {
        num = Number(tmp[0]);
        if (num < INT_MIN)
            return INT_MIN;
        if (num > INT_MAX)
            return INT_MAX;
    }

    return num;
};

File: 008/string-to-integer.py
class Solution:
    def myAtoi(self, str):
        """
        :type str: str
        :rtype: int
        """
        str = str.strip()
        num = 0

        INT_MIN = - 1 << 31
        INT_MAX = (1 << 31) - 1 # pay attention to its operation order

        import re
        tmp = re.findall(r'^[-+]?\d+', str)
        
        if tmp:
            num = int(tmp[0])
            if num < INT_MIN:
                return INT_MIN
            if num > INT_MAX:
                return INT_MAX

        return num


print(Solution().myAtoi("   -42"))
print(Solution().myAtoi("words and 987"))
print(Solution().myAtoi("+1"))
print(Solution().myAtoi("2147483646"))
